### PR TITLE
bump crengine: text selection and embedded fonts tweaks

### LINKED ISSUE #6217: When highlighting an epub koreader doesn't seem to include punctuation at end of highlighting
* KOReader version: 2020 05
* Device: pw4

#### Steps to reproduce

Also seems to apply to periods and other sentence ending punctuation

### LINKED ISSUE #8857: FR: highlight should include period / punctuation
When I try to highlight a complete sentence on my Onyx Boox device, the highlight typically stops before the period. If I bring up the highlight menu (to extend or contract the highlight by a word) this tool also skips the period. 

I’d like the period included with the highlight so that when I export my highlights to notes, I know that I have grabbed a complete sentence rather than a fragment.

### LINKED ISSUE #9169: Highlight menu does not show up when full stop is placed at line end
* KOReader version: 2022.05.1 (FDroid)
* Device: Onyx Boox Note 3

#### Issue
When trying to highlight a sentence/group of sentences that end with a full stop right at the rightmost edge of the screen, the highlight menu does not show up.

#### Steps to reproduce
1. Open an EPUB book (probably any reflowable)
2. Find a sentence that ends in a full stop right at the right margin, so that the next sentence starts in a new line\
3. Try highlighting that sentence up to and including the final full stop\
4. Get angry

#### Additional comments
Highlighting sentences with their corresponding punctuation at the end is very clunky and often takes multiple tries or requires editing the highlight via the provided buttons in edit mode. I see of no reason why KOReader shouldn't include the full stops automatically in highlights. Related issue: #4134, seems not to be solved fully.

### LINKED ISSUE #9190: FR: extract and show font name when Embedded Font is enabled
**Does your feature request involve difficulty completing a task? Please describe.**
That's a minor thing, but I think this might improve the UX quite a bit.

**Describe the solution you'd like**
In the `Font` menu, show that 1) the font is currently overridden by embedded fonts; 2) the font name of the font from the embedded style. 

**Describe alternatives you've considered**
The only KOReader-way to see this is by knowing that this setting is enabled, and knowing how to use HTML viewer and then to see the appropriate style css, then to find the font.

**Additional context**
I understand that enabling Embedded fonts most of times involves multiple font names. So maybe there should be at least mentioning that the fonts are overridden?

Use case is simply: to make KOReader even more user friendly when selecting fonts. 

P.S. Also I think that styles and fonts toggles should be in the same place with general fonts and styles. The same goes for Alt Status bar

### PR DESCRIPTION
**Pull request by poire-z**
#### bump crengine: text selection tweaks and helpers

Includes https://github.com/koreader/crengine/pull/484 : 
- lvtext: fix `m_kerning_mode` type
- XML: let '`gb2312`' (Chinese) encoding be known
- Add `ldomXPointer::getChar()`
- `LVDocView::getNodeByPoint()`: tweak for text selection
- Fonts: allow fetching the list of embedded fonts

cre.cpp https://github.com/koreader/koreader-base/pull/1490 : 
- `getTextFromPositions()`: use `getNodeByPoint(forTextSelection=true)` to allow panning in margins and get the nearest text instead of nothing.
Closes #9169.
- add `extendXPointersToSentenceSegment()` to allow extending some xpointer range to include punctuations at start or end (with for now a quite rudimentary implementation).
- add `getEmbeddedFontList()` to allow fetching the list of embedded fonts in the current EPUB book.

Also includes https://github.com/koreader/koreader-base/pull/1489 `Make luacheck >= 0.26 happy`.

#### Text highlighting: extend to include punctuations

When selected text seems to be a "sentence segment" (that is, when there are punctuations around start AND end), extend the selection to include the relevant punctuation.
Do this only when saving highlights and for translation (but not when dict or wikipedia lookups, or search).
See https://github.com/koreader/koreader/issues/8857#issuecomment-1151198923 for some screenshots showing what good (and bad) this does. Might need further refinements.
Closes #6217. Closes #8857.

#### Embedded fonts toggle: disabled if no embedded font

Have the enabled/disabled state of the toggle show the presence or not of embedded fonts in the current book.
Also show the names of the embedded fonts in the help_text InfoMessage.
See https://github.com/koreader/koreader/issues/9190#issuecomment-1152503765 for some screenshots. Any rewording suggestion ?
Closes #9190.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9198)

### CODE REVIEW

**Comment by poire-z** [comment]
Pinging @Frenzie for a validation/suggestion of the `Embedded fonts provided by the current book:` wording and the "not used" suffix and translators.
See https://github.com/koreader/koreader/issues/9190#issuecomment-1152503765 for screenshots.

**Comment by Frenzie** [comment]
Both sound good in principle. What does "not used" mean though, that the font is wasting disk space in the EPUB but isn't used?

File: frontend/ui/data/optionsutil.lua [hunk 1]
```diff
@@ -101,7 +101,10 @@ function optionsutil.showValues(configurable, option, prefix, document)
     end
     if option.help_text_func then
         -- Allow for concatenating a dynamic help_text_func to a static help_text
-        help_text = T("%1\n%2\n", help_text, option.help_text_func(configurable, document))
+        local more_text = option.help_text_func(configurable, document)
+        if more_text and more_text ~= "" then
+            help_text = T("%1\n%2\n", help_text, more_text)
```
Comment (Frenzie) [inline frontend/ui/data/optionsutil.lua:106]:
```suggestion
            help_text = T("%1\n%2", help_text, more_text) .. "\n"
```

Comment (poire-z) [reply frontend/ui/data/optionsutil.lua:106]: That thing is not translated (no `_()`), so it shouldn't really matter either way, right ?

**Comment by poire-z** [comment]
> Both sound good in principle. What does "not used" mean though, that the font is wasting disk space in the EPUB but isn't used?

Technically, that the font is not instantiated (it is instantiated when we request a specific font size to use it).
Which for a user, would mean either:
- what you say: it's in the EPUB, possibly referenced in some CSS, but no match, so not assigned to any node
- that you have a font of the same name that is used instead
- that you have disabled Embedded fonts :) so, ok, you know that, but this is a reminder

File: frontend/ui/data/optionsutil.lua (hunk 1 above)
Comment (Frenzie) [reply frontend/ui/data/optionsutil.lua:106]: Oops, my bad. I guess I was expecting string.format for that. :)

**Comment by offset-torque** [comment]
**Regarding: "Text highlighting: extend to include punctuations"**

Firstly thanks for this improvement poire-z.
It was really a necessary feature I think (for question sentences, post-export efforts etc.).

I want to report that this is not working in **Select and Highlight** mode which **hius07** implemented.
Also extending/shrinking a proper highlight reverts to the previous behavior it seems, it ignores punctutation as before.

Should I open an issue for these ?

**Comment by poire-z** [comment]
It works on Select and Highlight for me (select some initial word, Select, select some ending word, Highlight: the trailing punctuation is grabbed - dunno if there are other way to use this select&highlight).
As for the extend/shrink, I thought these were for manual tweaking, and actually correcting this auto-extend-and-grab-punctuation when it actually does not what you want for - so, dunno if it should really also happen there...

**Comment by offset-torque** [comment]
Ok I found the problem. 
I was reflexively trying to select the ending period too. This confuses KOReader I guess and it adds a word from the next sentence or paragraph. When I tap on the last word (instead of the last period) it properly adds the period itself. I think this needs a note in the user guide.

By the way I am not finding fault in the implementation, it is correct and better than the alternative. Tapping on a word is a much better experience than trying to tap on the last period (which always makes me fume on the stupid Android). Problem might be that users probably won't expect this smart behavior and try to select till the dot as they do on other operating systems. 

Maybe that **`Select ending fragment`** message can be replaced with **`Select ending word`** to indicate that a word should be tapped instead of the last character of the sentence.

**Regarding extend/shrink behavior**: If this is the intended behavior there is no problem.I agree that this behavior is the correct one. I just wanted to give feedback in the case that it is not intentional.

---

While testing I found out that if I start a highlight from the first word it adds the period correctly. 
But if I start from the second word it excludes the period (as in the image below). 
I thought the detection logic considers only the last word. 
Do we have to start from the first word if we want the ending punctuation ?

kobohighlightpunctutation

**Comment by mergen3107** [comment]
I think the idea was to be able to select a whole sentence, hence you need to select the first word.

However, I have though about an edge case: what if there is a monospaced lower-cased technical term in the very beginning? Or something else that can only be written in lower case, like company names / brands (calibre :D)

**Comment by poire-z** [comment]
> I think the idea was to be able to select a whole sentence, hence you need to select the first word.

Right. See my thinking at (or around): https://github.com/koreader/koreader/issues/8857#issuecomment-1151198923

> Do we have to start from the first word if we want the ending punctuation ?

Yes. I guess you would not want "weaknesses of markets." to be selected if you selected only "weaknesses of markets" ?

> However, I have though about an edge case: what if there is a monospaced lower-cased technical term in the very beginning? Or something else that can only be written in lower case, like company names / brands (calibre :D)

The code doesn't care about lowercase/uppercase. It's supposed to work with CJK, Arabic,... and languages with no lower/uppercase.
So, what it looks about is not the first word being uppercase, but the stuff before this first word being end of sentence punctuation.

> Regarding extend/shrink behavior: If this is the intended behavior there is no problem.I agree that this behavior is the correct one. I just wanted to give feedback in the case that it is not intentional.

Well, may be it should still work with "move by word" - but obviously not when "move by char".

**Comment by offset-torque** [comment]
Thanks for the answers.
Now I understand the reasoning and everything works correctly it seems. 
(I am also asking these kind of questions to be able to explain a feature properly in the guide)

**Comment by poire-z** [comment]
> Well, may be it should still work with "move by word" - but obviously not when "move by char".

This could be achieved with:
```diff
--- a/frontend/apps/reader/modules/readerhighlight.lua
+++ b/frontend/apps/reader/modules/readerhighlight.lua
@@ -685,6 +685,13 @@ function ReaderHighlight:updateHighlight(page, index, side, direction, move_by_c
             end
         end
     end
+    if not move_by_char then
+        local extended_text = self.ui.document:extendXPointersToSentenceSegment(self.view.highlight.saved[page][index].pos0, self.view.highlight.saved[page][index].pos1)
+        if extended_text then
+            self.view.highlight.saved[page][index].pos0 = extended_text.pos0
+            self.view.highlight.saved[page][index].pos1 = extended_text.pos1
+        end
+    end

     local new_beginning = self.view.highlight.saved[page][index].pos0
     local new_end = self.view.highlight.saved[page][index].pos1
```
But in some cases, that I think would not be uncommon, you get some strange feeling (although logical) that when moving one end (which may not movein some cases, at node boundaries), you'll see the other end move :/
selection

So, dunno :/

**Pull request merged**